src/relay/backend/contrib/cmsisnn/codegen_cmsisnn.cc

```diff
@@ -16,190 +16,36 @@
-
-  std::string GetSource(const std::string& format) final { return code_; }
-
-  const char* type_key() const { return "c"; }
-
-  PackedFunc GetFunction(const std::string& name, const ObjectPtr<Object>& sptr_to_self) final {
-    if (name == "get_symbol") {
-      return PackedFunc(
-          [sptr_to_self, this](TVMArgs args, TVMRetValue* rv) { *rv = this->func_names_[0]; });
-    } else if (name == "get_func_names") {
-      return PackedFunc(
-          [sptr_to_self, this](TVMArgs args, TVMRetValue* rv) { *rv = this->func_names_; });
-    } else {
-      return PackedFunc(nullptr);
-    }
-  }
-
-  void SaveToFile(const std::string& file_name, const std::string& format) final {
-    std::string fmt = GetFileFormat(file_name, format);
-    std::string meta_file = GetMetaFilePath(file_name);
-    if (fmt == "c" || fmt == "cu") {
-      ICHECK_NE(code_.length(), 0);
-      SaveBinaryToFile(file_name, code_);
-    } else {
-      ICHECK_EQ(fmt, fmt_) << "Can only save to format=" << fmt_;
-    }
-  }
-
- protected:
-  std::string code_;
-  std::string fmt_;
-  Array<String> func_names_;
-};
-
-class CMSISNNModule : public Module {
- public:
-  CMSISNNModule() {}
-  explicit CMSISNNModule(ObjectPtr<Object> n) : Module(n) {}
-  inline CMSISNNModuleNode* operator->();
-  inline const CMSISNNModuleNode* operator->() const;
-};
-
-inline CMSISNNModuleNode* CMSISNNModule::operator->() {
-  return static_cast<CMSISNNModuleNode*>(get_mutable());
-}
-
-static Module CMSISNNModuleNodeCreate(IRModule mod) {
-  bool output_ssa = false;
-  CodeGenCMSISNN cg;
-  Array<String> function_names;
-  cg.Init(output_ssa);
-  ICHECK(mod->functions.size() == 1) << "Supports modules with single PrimFunc.";
-  for (auto kv : mod->functions) {
-    ICHECK(kv.second->IsInstance<PrimFuncNode>()) << "CodegenCHost: Can only take PrimFunc";
-    auto f = Downcast<PrimFunc>(kv.second);
-    auto global_symbol = f->GetAttr<String>(tvm::attr::kGlobalSymbol);
-    ICHECK(global_symbol.defined())
-        << "CodeGenCHost: Expect PrimFunc to have the global_symbol attribute";
-    function_names.push_back(global_symbol.value());
-    cg.AddFunction(f);
-  }
-  std::string code = cg.Finish();
-  auto n = make_object<CMSISNNModuleNode>(code, "c", function_names);
-  return Module(n);
+namespace relay {
+namespace contrib {
+namespace cmsisnn {
+
+transform::Pass RelayToTIR();
+
+runtime::Module CompileCMSISNN(const ObjectRef& ref) {
+  IRModule relay_mod;
+  Function relay_func = Downcast<Function>(ref);
+  auto func_name = relay_func->GetAttr<String>(tvm::attr::kGlobalSymbol);
+  GlobalVar var = GlobalVar(func_name.value());
+  relay_mod->Add(var, relay_func);
+  relay_mod = transform::InferType()(relay_mod);
+
+  Array<transform::Pass> pass_seqs{transform::InferType(), RelayToTIR()};
+  transform::Sequential seq(pass_seqs);
+  IRModule tir_mod = seq(relay_mod);
+
+  const auto* pf = runtime::Registry::Get("runtime.CMSISNNModuleNodeCreate");
```

Comment:
Can we just reference this from the C++ namespace rather than using the function registry?

Reply:
I wanted to do that, Just didn't do it because the definition CMSISNNModuleNodeCreate is in cc file not in the header. This block of code inside codegen_cmsisnn.py should go away later with target hooks later on.

Reply:
If you don't want an additional header, you can put the signature in the calling file, similar to: https://github.com/apache/tvm/blob/main/src/relay/backend/te_compiler.cc#L51

Reply:
We could do that in the follow up. But I don't see the need of it in the long run. Just hoping the entire codegen_cmsisnn.cc is not required in future.

Reply:
As I said, happy to see it in a follow up :smile_cat: